system-parachains/common/src/randomness.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot. If not, see <http://www.gnu.org/licenses/>.
+
+use core::marker::PhantomData;
+use cumulus_primitives_core::relay_chain;
+use frame_support::traits::Randomness;
+use sp_runtime::traits::{BlakeTwo256, Hash};
+use sp_state_machine::{Backend, TrieBackendBuilder};
+
+pub const LOG_TARGET: &str = "runtime::randomness";
+
+/// VRF output length for per-slot randomness.
+pub const VRF_RANDOMNESS_LENGTH: usize = 32;
+
+// The initial version of the `Randomness` implementation proposed by @gui1117
+/// Provides randomness from the Relay Chain VRF from one epoch ago, but does not include the block
+/// number indicating when this randomness was generated or became observable to chain observers.
+///
+/// WARNING: This implementation does not return the block number associated with the randomness,
+/// because this information is not available in the validation data.
+pub struct RelayChainOneEpochAgoWithoutBlockNumber<T, BlockNumber>(PhantomData<(T, BlockNumber)>);
+
+impl<T, BlockNumber> Randomness<T::Hash, BlockNumber>
+	for RelayChainOneEpochAgoWithoutBlockNumber<T, BlockNumber>
+where
+	T: cumulus_pallet_parachain_system::Config,
+	BlockNumber: From<u32>,
+{
+	fn random(subject: &[u8]) -> (T::Hash, BlockNumber) {
+		// Defensive fallback used if the `well_known_keys::ONE_EPOCH_AGO_RANDOMNESS` key
+		// is missing or absent from the validation data. This situation is unexpected,
+		// as the key should always be present.
+		let defensive_fallback = || (T::Hashing::hash(subject), 0.into());
```

Comment:
Maybe the fallback can use the parachain parent state root? it would be safer IIRC it is easily accessible.